python3-pytest: add new package

ja-pa opened the pull request:
Maintainer: me  @ja-pa
Compile tested: Turris Omnia (TOS4), OpenWrt 18.06.2
Run tested: Turris Omnia (TOS4), OpenWrt 18.06.2

Description:
Pytest is a popular testing framework for python.  It allows to write scalable and complex test set. It also supports multiple useful plugins like xdist  pytest-django and more (see http://plugincompat.herokuapp.com)

I runtested package with examples provided in [docs](https://docs.pytest.org/en/latest/usage.html#detailed-summary-report)

This PR also adds python dependency required by pytest
- **pluggy**
- **atomicwrites** 
- **more-itertools**
- **py**
- **python3-importlib-metadata**
- **python3-wcwidth**
- **python3-zipp**
- **python3-packaging**
- **python3-pyparsing**

Why add this package:
* Pytest is a dependency for Deckard (https://gitlab.labs.nic.cz/knot/deckard/ )  utility, which tests DNS resolvers on the network to let users know when their ISP is mishandling DNS.  I would like to submit Deckard in some future PR.
* It's nice to have pytest on a router for testing new python packages and also updates.

jefferyto commented on lang/python/atomicwrites/Makefile:
```diff
@@ -0,0 +1,42 @@
+  CATEGORY:=Languages
+  TITLE:=atomicwrites
+  URL:=https://python-atomicwrites.readthedocs.io/en/latest/
+  DEPENDS:=+python3-ligh
```
I think this is causing the CI error.

ja-pa replied:
Thanks, yes that was the problem.

ja-pa commented:
cc @jefferyto, @commodo

jefferyto commented on lang/python/pytest/Makefile:
```diff
@@ -0,0 +1,49 @@
+	+python3-more-itertools \
+	+python3-py \
+	+python3-six \
+	+python3-pluggy
```
Looking at [`setup.py`](https://github.com/pytest-dev/pytest/blob/4.4.0/setup.py#L5-L17), it looks like pytest also needs attrs. Are you sure this package works without it?

ja-pa replied:
I decided to remove xdist from this PR

commodo commented on lang/python/pluggy/Makefile:
```diff
@@ -22,6 +22,9 @@ PKG_LICENSE_FILES:=LICENSE
 include ../python3-package.mk
 
+PYTHON3_PKG_SETUP_VARS:= \
+        PKG_VERSION="$(PKG_VERSION)"
```
is there a `PKG_VERSION` variable in `pluggy` ?

commodo replied:
ah; nvm; found it

commodo commented:
From my side, this looks good (generally).
After resolving comments on this and fixing the build, it should be ok.

Looks like some `os` imports need handling
```
Traceback (most recent call last):
  File "./setup.py", line 45, in <module>
    main()
  File "./setup.py", line 31, in main
    version=os.getenv('PKG_VERSION'),
NameError: name 'os' is not defined
```

jefferyto commented:
I see you have updated the `URL` fields - thanks!

***

Regarding setuptools-scm, I see you have patches to supply setuptools with version numbers, but this is only half the task. Taking pluggy as an example, we can see its [`use_scm_version` directive](https://github.com/pytest-dev/pluggy/blob/0.9.0/setup.py#L31) is:

```
use_scm_version={"write_to": "pluggy/_version.py"},
```

which tells setuptools-scm to both pass the version number to setuptools, and [write the version number](https://github.com/pypa/setuptools_scm#configuration-parameters) to the file "pluggy/_version.py".

The [template setuptools-scm uses for this file](https://github.com/pypa/setuptools_scm/blob/v3.2.0/src/setuptools_scm/__init__.py#L16-L21) looks like this:

```
# coding: utf-8
# file generated by setuptools_scm
# don't change, don't track in version control
version = {version!r}
```

Here `{version!r}` is basically replaced with the version number (technically a string). [pluggy tries to import this value](https://github.com/pytest-dev/pluggy/blob/0.9.0/pluggy/__init__.py#L1-L6) in its \_\_init\_\_.py:

```python
try:
    from ._version import version as __version__
except ImportError:
    # broken installation, we don't even try
    # unknown only works because we do poor mans version compare
__version__ = "unknown"
```

(`__version__` is a ["quasi-standard" variable](https://stackoverflow.com/a/459185) for Python packages to store/expose version numbers.) So in order to replicate this, we can adding something like this to the pluggy Makefile (somewhere after `include ../python3-package.mk`; note I haven't tested this):

```
define Py3Build/Compile
	echo "version = \"$(PKG_VERSION)\"" > $(PKG_BUILD_DIR)/pluggy/_version.py
	$(call Py3Build/Compile/Default)
endef
```

***

Would appreciate it if you can go through the affected packages (pluggy, py, pytest-xdist, pytest) and add this treatment (in addition to the "do not use setuptools-scm" patches). (I know the example I gave before (python-automat) didn't have this; that package didn't try to write the version number to a file.)

Thanks :pray: :joy:

ja-pa commented on lang/python/pytest-xdist/Makefile:
```diff
@@ -0,0 +1,41 @@
+  CATEGORY:=Languages
+  TITLE:=pytest-xdist
+  URL:=https://github.com/pytest-dev/pytest-xdist
+  DEPENDS:=+python3-pytest
```
I decided to remove xdist from this PR

ja-pa commented:
@jefferyto  @commodo  Hi, I believe that I fixed PR according to your suggestion. I removed pytest-xdist plugin package, updated all packages to the latest version and cherry-picked new one from https://github.com/openwrt/packages/pull/9096 . 
I run tested pytest with examples provided in docs and It worked fine. 
Will be glad for feedback, thx.

BKPepe commented on lang/python/python3-importlib-metadata/Makefile:
```diff
@@ -0,0 +1,38 @@
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=Like importlib_resources only for metadata.
```
Remove dot, please. It is not necessary here.

ja-pa replied:
removed

BKPepe commented on lang/python/python3-importlib-metadata/Makefile:
```diff
@@ -0,0 +1,38 @@
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=Like importlib_resources only for metadata.
+  URL:=https://gitlab.com/python-devs/importlib_metadata
```
@jefferyto What do you think that there should be Git repository instead of [documentation](https://importlib-metadata.readthedocs.io/en/latest/)? The URL for Documentation is listed actually as homepage on pypi.

BKPepe commented:
@ja-pa I have just one request. Would you please add in the commit message for python3-zipp and python3-importlib-metdata above your signed-off-by that you added patches for those two? We would have the same commit message format as OpenWrt in its main repository.

As an example what I mean is https://github.com/openwrt/openwrt/commit/64493d42b41253aafe1538e325e0692db4201046

It's up to you. It's not required, but it would be good to have it.

BKPepe commented on lang/python/python3-pyparsing/Makefile:
```diff
@@ -0,0 +1,45 @@
+endef
+
+define Package/python3-pyparsing/description
+  The pyparsing module is an alternative approach to creating and executing simple grammars,
```
Nitpick: in these two rows you have long indent and in the other two, you don't have it. Stick with two spaces as you have in the last two rows.

ja-pa replied:
fixed to two spaces

BKPepe commented:
Otherwise, it looks good to me.

ja-pa commented:
@BKPepe changed as requested. I added explanatory comments to python3-zipp and python3-importlib-metdata commit messages

aparcar commented:
@ja-pa You might like this tool here to automate pip package maintaining https://github.com/openwrt/packages/pull/9220

commodo commented:
One general note: maybe sending some of the `PKG_VERSION` patches upstream would be interesting.
It should be interesting for them as well as for us here.

aparcar commented:
@ja-pa There is no Copyright over the *cherry-picked* Makefiles, can this be a problem? If so please feel free to add your own Copyright.

ja-pa commented:
@aparcar For me it's no problem. If you what to give there your copyright info I have no problem with that.

aparcar commented:
@ja-pa are you trying to get you patches upstream? I'd be happy to get rid of these patches at some point. Apart from that I'd be happy to merge this (as I need pytest to work) - objections?

jefferyto commented:
I don't think the `PKG_VERSION` patches make sense upstream. I would assume they deliberately chose to use setuptools-scm.

Actually... now that I think about it, it should be possible to set:

```
HOST_PYTHON3_PACKAGE_BUILD_DEPENDS:=setuptools-scm
```

and have it installed host-side. I believe (although not 100% sure) all archives downloaded from PyPI have a [`PKG-INFO`](https://github.com/pypa/setuptools_scm#builtin-mechanisms-for-obtaining-version-numbers) with the version number, and if it doesn't, we can set [`SETUPTOOLS_SCM_PRETEND_VERSION`](https://github.com/pypa/setuptools_scm#environment-variables) in the Makefile.

Apologies @ja-pa for making you jump through all those hoops - while I think the current version number handling is good enough to be merged, this might be the easier/better way of managing this.

BKPepe commented on lang/python/python3-importlib-metadata/Makefile:
```diff
@@ -0,0 +1,38 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=importlib_metadata
+PKG_VERSION:=0.17
```
There is already version 0.18.

BKPepe commented on lang/python/python3-zipp/Makefile:
```diff
@@ -0,0 +1,37 @@
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=python3-zipp
+  URL:=http://pypi.python.org/zipp
```
```suggestion
  URL:=https://github.com/jaraco/zipp
```

BKPepe replied:
It's added. Thank you @ja-pa.
https://github.com/openwrt/packages/pull/8562/files/#diff-88499662378b6437f4a20610a1978130R23

BKPepe commented on lang/python/more-itertools/Makefile:
```diff
@@ -0,0 +1,43 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=more-itertools
+PKG_VERSION:=7.0.0
```
```suggestion
PKG_VERSION:=7.1.0
```

BKPepe commented on lang/python/more-itertools/Makefile:
```diff
@@ -0,0 +1,43 @@
+
+PKG_SOURCE:=more-itertools-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/m/more-itertools
+PKG_HASH:=c3e4748ba1aad8dba30a4886b0b1a2004f9a863837b8654e7059eebf727afa5a
```
```suggestion
PKG_HASH:=8bb43d1f51ecef60d81854af61a3a880555a14643691cc4b64a6ee269c78f09a
```

BKPepe commented on lang/python/more-itertools/Makefile:
```diff
@@ -0,0 +1,43 @@
+  SUBMENU:=Python
+  SECTION:=lang
+  CATEGORY:=Languages
+  TITLE:=more-itertools
```
```suggestion
  TITLE:=More routines for operating on iterables, beyond itertools
```

BKPepe commented on lang/python/pytest/Makefile:
```diff
@@ -0,0 +1,62 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=pytest
+PKG_VERSION:=4.6.2
```
```suggestion
PKG_VERSION:=5.0.0
```

BKPepe commented on lang/python/pytest/Makefile:
```diff
@@ -0,0 +1,62 @@
+
+PKG_SOURCE:=pytest-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/p/pytest/
+PKG_HASH:=bea27a646a3d74cbbcf8d3d4a06b2dfc336baf3dc2cc85cf70ad0157e73e8322
```
```suggestion
PKG_HASH:=95b700cf21ed5b7e91bce7a6b5a573b2e3ef7b3643d00f681d8f9c4672f9fbdf
```

BKPepe commented on lang/python/python3-importlib-metadata/Makefile:
```diff
@@ -0,0 +1,38 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=importlib_metadata
```
This is so wrong. It should have the same name as the folder, where is the Makefile.

BKPepe commented on lang/python/python3-importlib-metadata/Makefile:
```diff
@@ -0,0 +1,38 @@
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=Like importlib_resources only for metadata
```
```suggestion
  TITLE:=Read metadata from Python packages
```

BKPepe commented:
For PKG_NAME and as well the folder name see https://github.com/openwrt/packages/pull/8738#issuecomment-486241248

BKPepe commented on lang/python/python3-importlib-metadata/patches/001-fix-scm-version.patch:
```diff
@@ -0,0 +1,10 @@
+Index: importlib_metadata-0.17/setup.py
+===================================================================
+--- importlib_metadata-0.17.orig/setup.py
++++ importlib_metadata-0.17/setup.py
```
Why it doesn't look only for `setup.py` but it's looking for `importlib_metadata-0.17/setup.py`?

BKPepe replied:
Much better, thank you! :)

aparcar commented:
Wow this is a complicated PR, @ja-pa are you still motivated?

commodo commented:
i would just split this PR into multiple smaller PRs that adds some of these packages
3 months is not a good timeline for any PR

ja-pa commented:
@commodo Yes, looking back that would be a much better approach. But I think that right now it is best to continue in this PR since it's "almost" done.

BKPepe commented:
@ja-pa  would you please squash the latest commits, rebase it on current master and tell us if this is ready?

BKPepe commented on lang/python/python3-packaging/Makefile:
```diff
@@ -0,0 +1,42 @@
+
+PKG_MAINTAINER:=Jan Pavlinec <jan.pavlinec@nic.cz>
+PKG_LICENSE:=BSD-2-Clause Apache-2.0
+PKG_LICENSE_FILES:=LICENSE.APACHE LICENSE.BSD
```
```suggestion
PKG_LICENSE_FILES:=LICENSE LICENSE.APACHE LICENSE.BSD
```

ja-pa replied:
Added LICENSE file

BKPepe commented on lang/python/python3-wcwidth/Makefile:
```diff
@@ -0,0 +1,42 @@
+
+PKG_MAINTAINER:=Jan Pavlinec <jan.pavlinec@nic.cz>
+PKG_LICENSE:=MIT
+PKG_LICENSE_FILES:=LICENSE
```
I see it added https://github.com/openwrt/packages/pull/8562/files/#diff-fdc8f1b7c4faf43db6f06483529e7b44R21 , I'm marking it as resolved.

BKPepe commented on lang/python/python3-packaging/Makefile:
```diff
@@ -0,0 +1,42 @@
+PKG_HASH:=0c98a5d0be38ed775798ece1b9727178c4469d9c3b4ada66e8e6b7849f8732af
+
+PKG_MAINTAINER:=Jan Pavlinec <jan.pavlinec@nic.cz>
+PKG_LICENSE:=BSD
```
It's added. https://github.com/openwrt/packages/pull/8562/files/#diff-654d5e6a4fed9ee02775d3e690e2c166R20

I'm marking it as resolved.

BKPepe commented on lang/python/python3-importlib-metadata/Makefile:
```diff
@@ -0,0 +1,40 @@
+PKG_BUILD_DIR:=$(BUILD_DIR)/importlib_metadata-$(PKG_VERSION)
+
+PKG_MAINTAINER:=Paul Spooren <mail@aparcar.org>
+PKG_LICENSE:=Apache-2.0.
```
```suggestion
PKG_LICENSE:=Apache-2.0
```

ja-pa replied:
Fixed

BKPepe commented on lang/python/python3-zipp/Makefile:
```diff
@@ -0,0 +1,37 @@
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://pypi.python.org/packages/source/z/zipp/
```
@ja-pa  changed it.
https://github.com/openwrt/packages/pull/8562/files/#diff-88499662378b6437f4a20610a1978130R8

BKPepe commented on lang/python/python3-importlib-metadata/Makefile:
```diff
@@ -0,0 +1,40 @@
+  SUBMENU:=Python
+  TITLE:=Read metadata from Python packages
+  URL:=https://gitlab.com/python-devs/importlib_metadata
+  DEPENDS=+python3-light +pyhon3-zipp
```
```suggestion
  DEPENDS=+python3-light +python3-zipp
```

ja-pa replied:
Fixed

BKPepe commented on lang/python/pytest/patches/001-do-not-use-setuptools-scm.patch:
```diff
@@ -0,0 +1,21 @@
+Index: pytest-5.0.0/setup.py
+===================================================================
+--- pytest-5.0.0.orig/setup.py
++++ pytest-5.0.0/setup.py
```
Would you please remove it the exact version? If there is a new release, you will need to rebase the patch. There's a chance that you will don't need to do it for each version. In that case, the refresh should be enough.

ja-pa replied:
Removed

BKPepe commented on lang/python/python3-zipp/patches/001-fix-scm-version.patch:
```diff
@@ -0,0 +1,13 @@
+Index: zipp-0.5.1/setup.py
+===================================================================
+--- zipp-0.5.1.orig/setup.py
++++ zipp-0.5.1/setup.py
```
Would you please remove it the exact version? If there is a new release, you will need to rebase the patch. There's a chance that you will don't need to do it for each version. In that case, the refresh should be enough.

ja-pa replied:
Removed

BKPepe commented on lang/python/py/patches/001-do-not-use-setuptools.patch:
```diff
@@ -0,0 +1,19 @@
+Index: py-1.8.0/setup.py
+===================================================================
+--- py-1.8.0.orig/setup.py
++++ py-1.8.0/setup.py
```
Would you please remove it the exact version? If there is a new release, you will need to rebase the patch. There's a chance that you will don't need to do it for each version. In that case, the refresh should be enough.

ja-pa replied:
Removed

BKPepe commented on lang/python/pluggy/patches/001-do-not-use-setuptools-scm.patch:
```diff
@@ -0,0 +1,20 @@
+Index: pluggy-0.12.0/setup.py
+===================================================================
+--- pluggy-0.12.0.orig/setup.py
++++ pluggy-0.12.0/setup.py
```
Would you please remove it the exact version?

ja-pa replied:
Removed

ja-pa replied:
Removed

BKPepe commented on lang/python/atomicwrites/Makefile:
```diff
@@ -0,0 +1,41 @@
+  SUBMENU:=Python
+  SECTION:=lang
+  CATEGORY:=Languages
+  TITLE:=atomicwrites
```
```suggestion
  TITLE:=Atomic file writes
```

ja-pa replied:
changed

ja-pa replied:
Fixed

BKPepe commented on lang/python/pluggy/Makefile:
```diff
@@ -0,0 +1,49 @@
+  SUBMENU:=Python
+  SECTION:=lang
+  CATEGORY:=Languages
+  TITLE:=pluggy
```
```suggestion
  TITLE:=Plugin and hook calling mechanisms for Python
```

ja-pa replied:
fixed with longer title description

ja-pa replied:
Fixed

BKPepe commented on lang/python/pytest/Makefile:
```diff
@@ -0,0 +1,62 @@
+  SUBMENU:=Python
+  SECTION:=lang
+  CATEGORY:=Languages
+  TITLE:=pytest
```
```suggestion
  TITLE:=pytest: simple powerful testing with Python
```

ja-pa replied:
Fixed

BKPepe commented on lang/python/python3-zipp/Makefile:
```diff
@@ -0,0 +1,37 @@
+  SECTION:=lang
+  CATEGORY:=Languages
+  SUBMENU:=Python
+  TITLE:=python3-zipp
```
```suggestion
  TITLE:=Backport of pathlib-compatible object wrapper for zip files
```

Hopefully, it is not too long to be shown in `make menuconfig`.

ja-pa replied:
Fixed

ja-pa commented:
@commodo @jefferyto @BKPepe I think, that I solved all suggested issues. Could you review that? When it comes to HOST_PYTHON3_PACKAGE_BUILD_DEPENDS:=setuptools-scm, I would rather change it in another PR.

jefferyto commented on lang/python/pytest/Makefile:
```diff
@@ -0,0 +1,62 @@
+  URL:=https://docs.pytest.org/en/latest/
+  DEPENDS:= \
+	+python3-light \
+	+python3-setuptools \
```
[setuptools is not an install dependency](https://github.com/pytest-dev/pytest/blob/5.0.0/setup.py#L5-L16).

jefferyto commented on lang/python/pytest/Makefile:
```diff
@@ -0,0 +1,62 @@
+	+python3-atomicwrites \
+	+python3-more-itertools \
+	+python3-py \
+	+python3-six \
```
[six is no longer a dependency](https://github.com/pytest-dev/pytest/blob/5.0.0/setup.py#L5-L16).

jefferyto commented on lang/python/pytest/Makefile:
```diff
@@ -0,0 +1,62 @@
+	+python3-importlib-metadata \
+	+python3-packaging \
+	+python3-wcwidth \
+	+python3-zipp
```
[zipp is still not a dependency](https://github.com/pytest-dev/pytest/blob/5.0.0/setup.py#L5-L16).

jefferyto commented on lang/python/python3-packaging/Makefile:
```diff
@@ -0,0 +1,42 @@
+  SUBMENU:=Python
+  TITLE:=Core utilities for Python packages
+  URL:=https://github.com/pypa/packaging
+  DEPENDS:=+python3-light +python3-pyparsing +python3-six +python3-attrs
```
attrs is a dependency in master but [not for 19.0](https://github.com/pypa/packaging/blob/19.0/setup.py#L51).

jefferyto commented on lang/python/python3-zipp/Makefile:
```diff
@@ -0,0 +1,37 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=zipp
+PKG_VERSION:=0.5.1
```
0.5.2 was released a few days ago.

jefferyto commented on lang/python/python3-zipp/Makefile:
```diff
@@ -0,0 +1,37 @@
+include $(TOPDIR)/rules.mk
```
I think python3-zipp is an accidental duplicate of python-zipp?

BKPepe commented on lang/python/python-atomicwrites/Makefile:
```diff
@@ -0,0 +1,42 @@
+PKG_SOURCE:=atomicwrites-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/a/atomicwrites/
+PKG_HASH:=75a9445bac02d8d058d5e1fe689654ba5a6556a1dfd8ce6ec55a0ed79866cfa6
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(atomicwrites-$(PKG_VERSION)
```
```suggestion
PKG_BUILD_DIR:=$(BUILD_DIR)/atomicwrites-$(PKG_VERSION)
```

BKPepe commented on lang/python/python-more-itertools/Makefile:
```diff
@@ -0,0 +1,43 @@
+PKG_SOURCE:=more-itertools-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/m/more-itertools
+PKG_HASH:=8bb43d1f51ecef60d81854af61a3a880555a14643691cc4b64a6ee269c78f09a
+
```
```suggestion

PKG_BUILD_DIR:=$(BUILD_DIR)/more-itertools-$(PKG_VERSION)
```

BKPepe commented on lang/python/python-packaging/Makefile:
```diff
@@ -0,0 +1,42 @@
+PKG_SOURCE:=packaging-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/p/packaging/
+PKG_HASH:=0c98a5d0be38ed775798ece1b9727178c4469d9c3b4ada66e8e6b7849f8732af
+
```
```suggestion

PKG_BUILD_DIR:=$(BUILD_DIR)/packaging-$(PKG_VERSION)
```

BKPepe commented on lang/python/python-pluggy/Makefile:
```diff
@@ -0,0 +1,49 @@
+PKG_SOURCE:=pluggy-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/p/pluggy/
+PKG_HASH:=0825a152ac059776623854c1543d65a4ad408eb3d33ee114dff91e57ec6ae6fc
+
```
```suggestion

```suggestion

PKG_BUILD_DIR:=$(BUILD_DIR)/pluggy-$(PKG_VERSION)
```
```

BKPepe commented on lang/python/python-py/Makefile:
```diff
@@ -0,0 +1,49 @@
+PKG_SOURCE:=py-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/p/py/
+PKG_HASH:=dc639b046a6e2cff5bbe40194ad65936d6ba360b52b3c3fe1d08a82dd50b5e53
+
```
```suggestion
PKG_BUILD_DIR:=$(BUILD_DIR)/py-$(PKG_VERSION)

```

BKPepe commented on lang/python/python-pyparsing/Makefile:
```diff
@@ -0,0 +1,46 @@
+PKG_SOURCE:=pyparsing-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/p/pyparsing/
+PKG_HASH:=1873c03321fc118f4e9746baf201ff990ceb915f433f23b395f5580d1840cb2a
+
```
```suggestion
PKG_BUILD_DIR:=$(BUILD_DIR)/pyparsing-$(PKG_VERSION)

```

BKPepe commented on lang/python/python-pytest/Makefile:
```diff
@@ -0,0 +1,61 @@
+
+PKG_SOURCE:=pytest-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/p/pytest/
+PKG_HASH:=95b700cf21ed5b7e91bce7a6b5a573b2e3ef7b3643d00f681d8f9c4672f9fbdf
```
```suggestion
PKG_BUILD_DIR:=$(BUILD_DIR)/pytest-$(PKG_VERSION)

```

BKPepe commented on lang/python/python-pytest/Makefile:
```diff
@@ -0,0 +1,61 @@
+PKG_SOURCE:=pytest-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/p/pytest/
+PKG_HASH:=95b700cf21ed5b7e91bce7a6b5a573b2e3ef7b3643d00f681d8f9c4672f9fbdf
+
```
```suggestion
PKG_BUILD_DIR:=$(BUILD_DIR)/pytest-$(PKG_VERSION)

```

BKPepe commented on lang/python/python-wcwidth/Makefile:
```diff
@@ -0,0 +1,42 @@
+
+PKG_SOURCE:=wcwidth-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/w/wcwidth/
+PKG_HASH:=3df37372226d6e63e1b1e1eda15c594bca98a22d33a23832a90998faa96bc65e
```
```suggestion
PKG_BUILD_DIR:=$(BUILD_DIR)/wcwidth-$(PKG_VERSION)

```

BKPepe commented on lang/python/python-wcwidth/Makefile:
```diff
@@ -0,0 +1,42 @@
+PKG_SOURCE:=wcwidth-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/w/wcwidth/
+PKG_HASH:=3df37372226d6e63e1b1e1eda15c594bca98a22d33a23832a90998faa96bc65e
+
```
```suggestion
PKG_BUILD_DIR:=$(BUILD_DIR)/wcwidth-$(PKG_VERSION)

```

BKPepe commented on lang/python/python-zipp/Makefile:
```diff
@@ -0,0 +1,37 @@
+PKG_SOURCE:=zipp-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/z/zipp/
+PKG_HASH:=4970c3758f4e89a7857a973b1e2a5d75bcdc47794442f2e2dd4fe8e0466e809a
+
```
```suggestion
PKG_BUILD_DIR:=$(BUILD_DIR)/zipp-$(PKG_VERSION)

```

BKPepe commented on lang/python/python-importlib-metadata/Makefile:
```diff
@@ -0,0 +1,40 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=python-importlib_metadata
+PKG_VERSION:=0.18
```
```suggestion
PKG_VERSION:=0.19
```

BKPepe commented on lang/python/python-importlib-metadata/Makefile:
```diff
@@ -0,0 +1,40 @@
+
+PKG_SOURCE:=importlib_metadata-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://files.pythonhosted.org/packages/source/i/importlib_metadata
+PKG_HASH:=cb6ee23b46173539939964df59d3d72c3e0c1b5d54b84f1d8a7e912fe43612db
```
```suggestion
PKG_HASH:=23d3d873e008a513952355379d93cbcab874c58f4f034ff657c7a87422fa64e8
```

commodo commented:
this PR is kinda old now; since April

i don't know if we had a comment on this PR [but it was on one related to pytest], but at this point in time, it really looks like splitting this PR into smaller PRs is easier;
i know/remember the reply was "but we're almost there, so let's not split it just now"

at this rate, we will be stuck with this re-re-re-re-reviewing for another 4-5 months

also, this PR has 103 comments already;
let's take this as a learning opportunity and cut this PR here, and split [well, that's my 2 cents here]

also, i really-really-really-really-really don't want to offend anyone here; i am pretty relaxed now [since coming back from a vacation], but this reminds me of an old internet pic [which i find amusing in this case]:

[![reviews](https://miro.medium.com/max/510/1*QZcSJKl09TNCMFbq7UHksQ.png)](https://miro.medium.com/max/510/1*QZcSJKl09TNCMFbq7UHksQ.png)

ja-pa commented:
@commodo Yes, I think that it is best way how to handle this.

aparcar commented:
Maybe we should migrate this also to [pypi.mk](https://github.com/openwrt/packages/commit/635cb1309c4d52763c03bc00e7c367e06908b54f)?

ja-pa commented:
@aparcar Yes I agree (see MR #10405, #10407, #10408, #10410, #10411)

BKPepe commented:
ping @ja-pa

ja-pa commented:
Merged in this PR https://github.com/openwrt/packages/pull/8562 Thx all for the feedback!
